src/rewards/CreatorRewards.sol

```diff
@@ -0,0 +1,107 @@
+
+    error INSUFFICIENT_ETH();
+}
+
+abstract contract CreatorRewards is ICreatorRewards {
+    uint256 internal constant TOTAL_REWARD_PER_MINT = 0.000999 ether;
+
+    uint256 internal constant CREATOR_REWARD_FREE_MINT = 0.000555 ether;
+    uint256 internal constant FINDER_REWARD_FREE_MINT = 0.000111 ether;
+    uint256 internal constant LISTER_REWARD_FREE_MINT = 0.000111 ether;
+    uint256 internal constant ZORA_REWARD_FREE_MINT = 0.000222 ether;
+
+    uint256 internal constant FINDER_REWARD_PAID_MINT = 0.000333 ether;
+    uint256 internal constant LISTER_REWARD_PAID_MINT = 0.000333 ether;
+    uint256 internal constant ZORA_REWARD_PAID_MINT = 0.000333 ether;
+
+    address internal immutable ZORA_REWARD_RECIPIENT;
+
+    IRewardsManager internal immutable REWARDS_MANAGER;
+
+    constructor(address _rewardsManager, address _zoraRewardRecipient) {
+        REWARDS_MANAGER = IRewardsManager(_rewardsManager);
+        ZORA_REWARD_RECIPIENT = _zoraRewardRecipient;
+    }
+
+    function _parseAttachedEth(uint256 msgValue, uint256 numTokens) internal pure returns (uint256) {
+        uint256 minEthRequired = numTokens * TOTAL_REWARD_PER_MINT;
+
+        if (msgValue < minEthRequired) {
+            revert INSUFFICIENT_ETH();
+        } else if (msgValue == minEthRequired) {
+            return 0;
+        } else {
+            unchecked {
+                return msgValue - minEthRequired;
+            }
+        }
+    }
+
+    function _transferFreeMintRewards(uint256 numTokens, address creator, address finder, address lister) internal {
+        (uint256 creatorReward, uint256 zoraReward, uint256 finderReward, uint256 listerReward, uint256 totalReward) = _computeFreeMintRewards(numTokens);
+
+        if (finder == address(0)) {
+            finder = ZORA_REWARD_RECIPIENT;
+        }
+
+        if (lister == address(0)) {
+            lister = ZORA_REWARD_RECIPIENT;
+        }
+
+        REWARDS_MANAGER.addReward{value: totalReward}(creator, creatorReward, ZORA_REWARD_RECIPIENT, zoraReward, finder, finderReward, lister, listerReward);
+
+        emit FreeMintRewardsTransferred(creator, creatorReward, finder, finderReward, lister, listerReward);
+    }
+
+    function _transferPaidMintRewards(uint256 msgValue, uint256 numTokens, address finder, address lister) internal returns (uint256 remainingEth) {
+        (uint256 zoraReward, uint256 finderReward, uint256 listerReward, uint256 totalReward) = _computePaidMintRewards(numTokens);
+
+        remainingEth = msgValue - totalReward;
+
+        if (finder == address(0)) {
+            finder = ZORA_REWARD_RECIPIENT;
+        }
+
+        if (lister == address(0)) {
+            lister = ZORA_REWARD_RECIPIENT;
+        }
+
+        REWARDS_MANAGER.addReward{value: totalReward}(ZORA_REWARD_RECIPIENT, zoraReward, finder, finderReward, lister, listerReward);
+
+        emit PaidMintRewardsTransferred(finder, finderReward, lister, listerReward);
+    }
+
+    function _computeFreeMintRewards(
```

Comment:
would be good to have an overload that computes this when there is no finder or lister